From 2f2150fb247a6deb3bfdd3ac3887441861e6936c Mon Sep 17 00:00:00 2001
From: Sathishkumar Muruganandam <quic_murugana@quicinc.com>
Date: Thu, 18 May 2023 15:15:43 +0530
Subject: [PATCH] WPS: fix multi-ap fronthaul association

Setting 4addr mode on fronthaul BSS breaks WPS association on backhaul STA
which is still on 3addr mode.

Fix the interop issue by not setting 4addr mode on fronthaul BSS with WPS
instead of setting by default for multi-ap case.

Fronthaul BSS for non-WPS STA shall continue to use 4addr mode.

Signed-off-by: Sathishkumar Muruganandam <quic_murugana@quicinc.com>
---
 src/ap/ieee802_11.c | 5 ++++-
 src/ap/sta_info.c   | 4 +++-
 2 files changed, 7 insertions(+), 2 deletions(-)

diff --git a/src/ap/ieee802_11.c b/src/ap/ieee802_11.c
index e3e643bbf776..007f3f9d0d49 100644
--- a/src/ap/ieee802_11.c
+++ b/src/ap/ieee802_11.c
@@ -7395,7 +7395,10 @@ static void handle_assoc_cb(struct hostapd_data *hapd,
 		sta->flags |= WLAN_STA_WDS;
 	}
 
-	if (sta->flags & (WLAN_STA_WDS | WLAN_STA_MULTI_AP)) {
+	/* WPS not supported on backhaul BSS & disable 4addr mode on FH */
+	if (sta->flags & WLAN_STA_WDS || (sta->flags & WLAN_STA_MULTI_AP &&
+		!(hapd->conf->multi_ap & FRONTHAUL_BSS) &&
+		!(sta->flags & WLAN_STA_WPS))) {
 		int ret;
 		char ifname_wds[IFNAMSIZ + 1];
 
diff --git a/src/ap/sta_info.c b/src/ap/sta_info.c
index f9e023d238ca..6d15a01cd58a 100644
--- a/src/ap/sta_info.c
+++ b/src/ap/sta_info.c
@@ -276,7 +276,9 @@ void ap_free_sta(struct hostapd_data *hapd, struct sta_info *sta)
 	ap_sta_set_authorized(hapd, sta, 0);
 	hostapd_set_sta_flags(hapd, sta);
 
-	if (sta->flags & (WLAN_STA_WDS | WLAN_STA_MULTI_AP))
+	if (sta->flags & WLAN_STA_WDS || (sta->flags & WLAN_STA_MULTI_AP &&
+		!(hapd->conf->multi_ap & FRONTHAUL_BSS) &&
+		!(sta->flags & WLAN_STA_WPS)))
 		hostapd_set_wds_sta(hapd, NULL, sta->addr, sta->aid, 0);
 
 	if (sta->ipaddr)
-- 
2.17.1

